kvbc/include/blockchain_view.h

```diff
@@ -0,0 +1,274 @@
+//
+// This product may include a number of subcomponents with separate copyright
+// notices and license terms. Your use of these subcomponents is subject to the
+// terms and conditions of the sub-component's license, as noted in the
+// LICENSE file.
+
+#pragma once
+
+#include <cstddef>
+#include <cstdint>
+#include <exception>
+#include <iterator>
+#include <limits>
+#include <type_traits>
+#include <unordered_map>
+#include <utility>
+
+#include <boost/iterator/iterator_categories.hpp>
+#include <boost/iterator/iterator_facade.hpp>
+
+#include "kv_types.hpp"
+
+namespace concord {
+namespace kvbc {
+
+namespace detail {
+
+template <typename BlockInfo>
+using BlockInfoCache = std::unordered_map<concordUtils::BlockId, BlockInfo>;
+
+using BlockIdDifferenceType = std::make_signed<concordUtils::BlockId>::type;
+
+}  // namespace detail
+
+// The maximum possible number of blocks this view is able to hold.
+inline constexpr auto MAX_BLOCKCHAIN_VIEW_SIZE = std::numeric_limits<detail::BlockIdDifferenceType>::max();
+
+// A random access blockchain iterator that supports caching block info.
+template <typename BlockInfo, bool loadDataOnAccess>
+class BlockchainIterator : public boost::iterator_facade<BlockchainIterator<BlockInfo, loadDataOnAccess>,
+                                                         const BlockInfo,  // The value type this iterator points to.
+                                                         boost::random_access_traversal_tag,
+                                                         const BlockInfo&,  // Reference to the value type.
+                                                         detail::BlockIdDifferenceType> {
+ private:
+  friend class boost::iterator_core_access;
+  template <typename, bool>
+  friend class BlockchainView;
+
+  using Cache = detail::BlockInfoCache<BlockInfo>;
+
+  BlockchainIterator(concordUtils::BlockId id,
+                     concordUtils::BlockId genesisId,
+                     concordUtils::BlockId endId,
+                     Cache& cache)
+      : blockId_{id}, genesisId_{genesisId}, endId_{endId}, cache_{&cache} {
+    cacheBlock(blockId_);
+  }
+
+  void cacheBlock(concordUtils::BlockId id) const {
+    if (isOutsideRange(id)) {
+      return;
+    }
+
+    // already cached, don't do any work
+    if (auto it = cache_->find(id); it != std::cend(*cache_)) {
+      return;
+    }
+
+    // load data from storage and, if successful, add to cache
+    auto blockInfo = BlockInfo{id};
+    blockInfo.loadIndices();
+    if constexpr (loadDataOnAccess) {
+      blockInfo.loadData();
+    }
+    cache_->emplace(id, std::move(blockInfo));
+  }
+
+  void increment() {
+    cacheBlock(blockId_ + 1);
+    ++blockId_;
+  }
+
+  void decrement() {
+    cacheBlock(blockId_ - 1);
+    --blockId_;
+  }
+
+  bool equal(const BlockchainIterator& other) const noexcept { return blockId_ == other.blockId_; }
+
+  const BlockInfo& dereference() const { return cache_->find(blockId_)->second; }
```

Comment:
Since we always insert into the cache when constructing or incrementing/decrementing `BlockchainIterator` objects, we expect find will not return `cend()`. Dereferencing `BlockchainIterator::cend()` is defined as undefined behaviour in itself. Additionally, we don't expect boost calling `dereference()`.

Good point, I will document that.